### PR TITLE
Fixed #113: PRINS-3998: Add support for ServiceComm.NET 0.7

### LINKED ISSUE #113: Add support for ServiceComm.NET 0.7

### PR DESCRIPTION
**Pull request by pekueble**
ServerRunner and ClientConsole work together (Not implemented exception), however there are heartbeat-exceptions logged.

### CODE REVIEW

File: com.btc.serviceidl/src/com/btc/serviceidl/generator/dotnet/DotNetGenerator.xtend [hunk 1]
```diff
@@ -90,6 +90,8 @@ class DotNetGenerator
         this.qualifiedNameProvider = qualifiedNameProvider
         this.generationSettings = generationSettings
         this.protobufProjectReferences = protobufProjectReferences
+        this.nugetPackages = new NuGetPackageResolver(ServiceCommVersion.get(generationSettings.getTargetVersion(
```
Comment (sigiesec) [inline com.btc.serviceidl/src/com/btc/serviceidl/generator/dotnet/DotNetGenerator.xtend]: I don't think this is necessary in the constructor. This is always overwritten within reinitializeProject

File: com.btc.serviceidl/src/com/btc/serviceidl/generator/dotnet/NuGetPackageResolver.xtend [hunk 2]
```diff
@@ -35,6 +42,19 @@ class NuGetPackageResolver
         "Common.Logging" -> "1.2.0"
     }
 
+    // ServiceComm 0.7
```
Comment (sigiesec) [inline com.btc.serviceidl/src/com/btc/serviceidl/generator/dotnet/NuGetPackageResolver.xtend]: This is widely a duplicate of versionMapper_0_6, the only difference is the BTC.CAB.ServiceComm.NET version. Please remove the duplication by only specifying the common dependencies once.

File: com.btc.serviceidl/src/com/btc/serviceidl/generator/dotnet/NuGetPackageResolver.xtend [hunk 3]
```diff
@@ -62,11 +82,21 @@ class NuGetPackageResolver
 
     val nugetPackages = new HashSet<NuGetPackage>
 
-    private static def NuGetPackage resolvePackageInternal(String assemblyName)
+    private def NuGetPackage resolvePackageInternal(String assemblyName)
     {
-        val versions = validOrThrow(packageMapper.get(assemblyName), assemblyName, "package ID").map [
-            new Pair(it, validOrThrow(versionMapper.get(it), it, "package version"))
-        ].toList
+        var List<Pair<String, String>> versions = null
```
Comment (sigiesec) [inline com.btc.serviceidl/src/com/btc/serviceidl/generator/dotnet/NuGetPackageResolver.xtend]: Revert this to a val, e.g. by extracted the assigments below into a function with two return statements, which is called here.

File: com.btc.serviceidl/src/com/btc/serviceidl/generator/dotnet/NuGetPackageResolver.xtend [hunk 4]
```diff
@@ -62,11 +82,21 @@ class NuGetPackageResolver
...
+        if (serviceCommTargetVersion == ServiceCommVersion.V0_6)
+        {
+            versions = validOrThrow(packageMapper.get(assemblyName), assemblyName, "package ID").map [
```
Comment (sigiesec) [inline com.btc.serviceidl/src/com/btc/serviceidl/generator/dotnet/NuGetPackageResolver.xtend]: The only difference between the two cases is the versionMapper variable used, but the whole statement is duplicate. Remove this duplication.

File: com.btc.serviceidl/src/com/btc/serviceidl/generator/dotnet/ServiceCommVersion.java [hunk 5]
```diff
@@ -11,7 +11,8 @@
 package com.btc.serviceidl.generator.dotnet;
 
 public enum ServiceCommVersion {
-    V0_6("0.6");
+    V0_6("0.6"),
+    V0_7("0.7");
```
Comment (sigiesec) [inline com.btc.serviceidl/src/com/btc/serviceidl/generator/dotnet/ServiceCommVersion.java]: the default version should also be changed to 0.7

File: com.btc.serviceidl/src/com/btc/serviceidl/generator/dotnet/AppConfigGenerator.xtend [hunk 6]
```diff
@@ -36,6 +36,8 @@ class AppConfigGenerator {
   // on Jenkins (but not on travis-ci)
   def CharSequence generateAppConfig(ModuleDeclaration module)
   {
+      val scv_V0_6 = getTargetVersion() == ServiceCommVersion.V0_6
```
Comment (sigiesec) [inline com.btc.serviceidl/src/com/btc/serviceidl/generator/dotnet/AppConfigGenerator.xtend]: do not use non-obvious abbreviations for variable names
maybe it does not make sense to extract this expression into a variable at all

File: com.btc.serviceidl/src/com/btc/serviceidl/generator/dotnet/NuGetPackageResolver.xtend [hunk 7]
```diff
@@ -16,14 +16,21 @@
 package com.btc.serviceidl.generator.dotnet
 
 import java.util.HashSet
+import java.util.List
 
 class NuGetPackageResolver
 {
+    ServiceCommVersion serviceCommTargetVersion
+    
+    new(ServiceCommVersion serviceCommTargetVersion){
```
Comment (sigiesec) [inline com.btc.serviceidl/src/com/btc/serviceidl/generator/dotnet/NuGetPackageResolver.xtend]: Remove the explicit constructor and add an `@Accessors(NONE)` annotation to the class, which automatically generates an equivalent constructor

File: com.btc.serviceidl/src/com/btc/serviceidl/generator/dotnet/ProxyGenerator.xtend [hunk 8]
```diff
@@ -34,6 +34,7 @@ class ProxyGenerator extends ProxyDispatcherGeneratorBase {
       if (featureProfile.usesEvents)
          resolve("BTC.CAB.ServiceComm.NET.Util.EventRegistryExtensions")
       val serviceFaultHandler = "serviceFaultHandler"
+      val scv_V0_6 = getTargetVersion() == ServiceCommVersion.V0_6
```
Comment (sigiesec) [inline com.btc.serviceidl/src/com/btc/serviceidl/generator/dotnet/ProxyGenerator.xtend]: analogous to AppConfigGenerator.xtend: rename or inline the variable

File: com.btc.serviceidl/src/com/btc/serviceidl/generator/dotnet/TestGenerator.xtend [hunk 9]
```diff
@@ -28,11 +28,14 @@ class TestGenerator extends GeneratorBase
     {
         val apiClassName = resolve(interfaceDeclaration)
         val loggerFactory = resolve("BTC.CAB.Logging.Log4NET.Log4NETLoggerFactory")
+        val serviceFaultHandlerManagerFactory = resolve("BTC.CAB.ServiceComm.NET.FaultHandling.ServiceFaultHandlerManagerFactory")
         val serverRegistration = getServerRegistrationName(interfaceDeclaration)
         val connectionFactory = resolve("BTC.CAB.ServiceComm.NET.SingleQueue.ZeroMQ.NetMQ.NetMqConnectionFactory")
+        val scv_V0_6 = getTargetVersion() == ServiceCommVersion.V0_6
```
Comment (sigiesec) [inline com.btc.serviceidl/src/com/btc/serviceidl/generator/dotnet/TestGenerator.xtend]: analogous to AppConfigGenerator.xtend: rename or inline the variable

**Comment by pekueble** [comment]
Did requested changes

**Comment by pekueble** [comment]
Did requested changes

**Comment by sigiesec** [comment]
@pekueble Most of the integration tests work, but the "event" test case is failing (also the BTC.GA.* test cases, but these probably fail for the same reason). The IDL file for the test case is at https://github.com/btc-ag/service-idl/blob/master/com.btc.serviceidl.tests/src/com/btc/serviceidl/tests/testdata/good/event.idl 

Could you please give this test case a look and fix the generated code for this as well?

Details on the error can be found in https://bpaste.net/show/6128a9067054, the first syntax error is mentioned in line 725.

**Comment by pekueble** [comment]
@sigiesec: using event.idl I could reproduce and fix this error.

**Comment by sigiesec** [comment]
Thanks! The integration tests are all fine now :)

**Pull request merged**